flocker/common/algebraic.py

```diff
@@ -0,0 +1,133 @@
+
+   class Dataset(PClass):
+       state = field(mandatory=True)
+       dataset_id = field(type=UUID, mandatory=True)
+       device_path = field(FilePath)
+       mount_point = field(FilePath)
+
+       __invariant__ = TaggedUnionInvariant(
+           tag_attribute='state',
+           attributes_for_tag={
+               DatasetStates.DETACHED: set(),
+               DatasetStates.ATTACHED_NO_FILESYSTEM: {'device_path'},
+               DatasetStates.ATTACHED: {'device_path'},
+               DatasetStates.MOUNTED: {'device_path', 'mount_point'},
+           },
+       )
+"""
+
+from pyrsistent import PClass, field
+
+from hypothesis.strategies import sampled_from, fixed_dictionaries, just
+
+__all__ = ["TaggedUnionInvariant"]
+
+
+class TaggedUnionInvariant(PClass):
+    """
+    An invariant that ensure the given object has an allowd tag
+    attribute, and that all the other specified attributes are
+    present if and only if the object has the appropriate tag.
+
+    Note: Attributes that aren't specified by any tag are ignored.
+
+    :param str tag_attribute: The attribute that contains the tag.
+    :param dict attributes_for_tag: Dictionary mapping tags to the
+        set of attributes allowed by that tag.
+    """
+
+    tag_attribute = field(str, mandatory=True)
+    attributes_for_tag = field(dict, mandatory=True)
+
+    @property
+    def _allowed_tags(self):
+        """
+        The set of all allowed tags.
+        """
+        return set(self.attributes_for_tag.keys())
+
+    @property
+    def _all_attributes(self):
+        """
+        The set of all attributes controlled by the invariant.
+        """
+        return {
+            attribute
+            for tag, attributes in self.attributes_for_tag.items()
+            for attribute in attributes
+        }
+
+    def __call__(self, value):
+        tag = getattr(value, self.tag_attribute)
+        if tag not in self._allowed_tags:
+            return (False, "can only be in {tag_name}s {tags}.".format(
+                tag_name=self.tag_attribute,
+                tags=', '.join(map("`{0.name}`".format,
+                                   self._allowed_tags)),
+            ))
+        for attribute in self.attributes_for_tag[tag]:
+            if not hasattr(value, attribute):
+                return (
+                    False,
+                    "`{attr}` must be specified in {tag_name} `{tag}`"
+                    .format(attr=attribute,
+                            tag_name=self.tag_attribute,
+                            tag=tag.name))
+        for attribute in self._all_attributes - self.attributes_for_tag[tag]:
+            if hasattr(value, attribute):
+                return (
+                    False,
+                    "`{attr}` can't be specified in {tag_name} `{tag}`"
+                    .format(attr=attribute,
+                            tag_name=self.tag_attribute,
+                            tag=tag.name))
+        return (True, "")
```

Comment:
:question: I guess if we could get a reliable list of fields that have been set on `value` then we could do this in one loop, iterating over those fields rather than the schema. I don't see a way to do this though.